### PR TITLE
Filon integration

### PR DESCRIPTION
**Pull request by regislebrun**
This method allows to integrate oscillatory functions such as characteristic functions of distributions.

### CODE REVIEW

File: python/src/FilonQuadrature_doc.i.in [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+    w(\omega t)=\left\{
+    \begin{array}{ll}
+      \cos(\omega t) & \mathrm{if}\: kind=0 \\
+      \sin(\omega t) & \mathrm{if}\: kind=1 \\
+      \exp(i \omega t) & \mathrm{if}\: kind\geq 2
+    \end{array}
+    \right.
+
+This algorithm is based on a regular partition of the interval :math:`[a,b]`, the function :math:`f` being approximated by a quadratic function on three consecutive points. This algorithm provides an approximation of order :math:`\cO(1/\omega^2)` when :math:`\omega\rightarrow\infty`. When :math:`w(\omega t)=\exp(i \omega t)`, the result is returned as a :class:`~openturns.Point` of dimension 2, the frist component being the real part of the result and the second component the imaginary part.
```
Comment (jschueller) [inline python/src/FilonQuadrature_doc.i.in]: please avoid lines too long, it's not easy to read

File: python/src/FilonQuadrature_doc.i.in [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    The integrand function.
+omega : float
+    The pulsation in the weight function. This value superseeds the value given in the constructor.
+interval : :class:`~openturns.Interval`, :math:`interval \in \Rset` 
+    The integration domain. 
+
+Returns
+-------
+value : :class:`~openturns.Point`
+    Approximation of the integral. Its dimension is :math:`p` if :math:`kind\in\{0,1\}`, otherwise it is :math:`2p` whith the :math:`p` first components corresponding to the real part of the integral and the remaining ones to the imaginary part.
```
Comment (jschueller) [inline python/src/FilonQuadrature_doc.i.in]: here too

**Comment by jschueller** [comment]
There is no test file.

**Comment by regislebrun** [comment]
It should now be ok. Any additional comments?

**Comment by jschueller** [comment]
Nope, autobuilder is happy :)

**Comment by regislebrun** [comment]
@dbarbier No, increasing n will NOT give you extra precision (at least: not in a systematic way). The error analysis of this kind of methods is quite counter-intuitive, the main point is that the relative error tends to zero when omega goes to infinity. The parameter n has an influence on the constant, but not in a straightforward way. I don't have the reference at hand right now but will give it to you ASAP. To check that everything is ok, it is better to fix n and to see what's append when omega grows.

**Comment by dbarbier** [comment]
No, there is a bug!
Replace
```
    c2n = f[0] * (0.5 * std::cos(x(0, 0)));
```
by
```
    c2n = f[0] * (0.5 * std::cos(omega * x(0, 0)));
```

**Comment by dbarbier** [comment]
BTW with polynomial functions I am pretty sure that increasing n does increase precision.  With the fix above, everything seems to work fine.

**Comment by regislebrun** [comment]
Damn! Another bug in the wall! Thanks for catching it. I spent some time with Sofiane talking about this method today, changed the test case and slightly reworked the implementation. I hope I didn't introduced the bug at this occasion.
Nevertheless, this method has a quite strange behavior. The discretization n can be decreased when omega increase while maintaining a reasonably good relative error!
An extension is to use Chebychev points instead of a regular grid as in [this paper](http://people.bath.ac.uk/masigg/publications/preprints/CleCur0810_final.pdf), but I didn't found the time to implement it. It would give a consistent high accuracy for the whole range of omega.

**Comment by dbarbier** [comment]
No, this bug is here since the beginning :-D

**Comment by regislebrun** [comment]
The same bug was also in the sine part...

**Comment by dbarbier** [comment]
And the last one:
```
-    valueSin = ((f[0] * std::cos(omega * x(0, 0)) - f[size - 1] * std::cos(omega * x(size - 1, 0))) * alpha - s2n * beta - s2nm1 * gamma) * h;
+    valueSin = ((f[0] * std::cos(omega * x(0, 0)) - f[size - 1] * std::cos(omega * x(size - 1, 0))) * alpha + s2n * beta + s2nm1 * gamma) * h;
```
Looks good after applying this fix.

**Pull request merged**